shared/util.go

```diff
@@ -843,6 +843,24 @@ func RunCommand(name string, arg ...string) (string, error) {
 	return string(output), nil
 }
 
+func RunCommandGetPid(name string, arg ...string) (int, error) {
+	cmd := exec.Command(name, arg...)
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	err := cmd.Start()
+	go cmd.Wait()
+	if err != nil {
+		err := RunError{
+			msg: fmt.Sprintf("Failed to run: %s %s", name, strings.Join(arg, " ")),
+			Err: err,
+		}
+		return -1, err
+	}
+
+	processPid := cmd.Process.Pid
+	return processPid, nil
+}
```

Comment:
If this is based on RunCommand (which it sounds like it is), it would make more sense to:
 - Move this to shared/util.go
 - Rebase RunCommand on it, so that RunCommand becomes a wrapper around RunCommandGetPid but ignores the returned pid.

Reply:
Doh, it's already in util.go :)
The second part still applies though.

Reply:
I am not exactly sure if there is a clean way to wrap RunCommandGetPid with RunCommand since RunCommand needs to wait for the process to finish and return its output. RunCommandGetPid should probably be renamed to something else since it essentially backgrounds the process which is not what RunCommand does. If it were to be wrapped by RunCommand, this is what I think it would look like:

```
func RunCommand(name string, arg ...string) (string, error) {
	var output bytes.Buffer
	cmd, pid, err := runCommandGetPid(&output, name, arg...)
	if err != nil {
		err := RunError{
			msg: fmt.Sprintf("Failed to run: %s %s: %s", name, strings.Join(arg, " "), strings.TrimSpace(output.String())),
			Err: err,
		}
		return output.String(), err
	}
	cmd.Wait()

	return output.String(), nil
}

func RunCommandGetPid(output *bytes.Buffer, name string, arg ...string) (*exec.Cmd, int, error) {
	cmd := exec.Command(name, arg...)
	cmd.Stdout = output
	cmd.Stderr = output	

	err := cmd.Start()
	if err != nil {
		return err
	}

	processPid := cmd.Process.Pid
	return cmd, processPid, nil
}
```
In my opinion, this does not look very clean because they are both named RunCommand but the wrapper waits for the process to finish and the wrapped function does not. I was thinking that RunCommandGetPid could be renamed to RunBackgroundCommand. Then it would make sense for it to return the PID of the background process because the caller would need to keep track it.

Reply:
SpawnCommand?

Reply:
```diff
diff --git a/shared/util.go b/shared/util.go
index e039905f..c178391b 100644
--- a/shared/util.go
+++ b/shared/util.go
@@ -830,17 +830,36 @@ func (e RunError) Error() string {
        return e.msg
 }
 
+func SpawnCommand(output *bytes.Buffer, name string, arg ...string) (*exec.Cmd, error) {
+       cmd := exec.Command(name, arg...)
+       cmd.Stdout = output
+       cmd.Stderr = output
+
+       err := cmd.Start()
+       if err != nil {
+               return nil, err
+       }
+
+       return cmd, nil
+}
+
 func RunCommand(name string, arg ...string) (string, error) {
-       output, err := exec.Command(name, arg...).CombinedOutput()
+       output := bytes.Buffer{}
+       cmd, err := SpawnCommand(&output, name, arg...)
+       if err != nil {
+               return "", err
+       }
+
+       err = cmd.Wait()
        if err != nil {
                err := RunError{
-                       msg: fmt.Sprintf("Failed to run: %s %s: %s", name, strings.Join(arg, " "), strings.TrimSpace(string(output))),
+                       msg: fmt.Sprintf("Failed to run: %s %s: %s", name, strings.Join(arg, " "), strings.TrimSpace(output.String())),
                        Err: err,
                }
-               return string(output), err
+               return output.String(), err
        }
 
-       return string(output), nil
+       return output.String(), nil
 }
 
 func TryRunCommand(name string, arg ...string) (string, error) {
```

Reply:
Since SpawnCommand returns exec.Cmd, we don't really need to return the pid separately, the caller can grab it easily enough with .Process.PID